gazelle/sass.go

```diff
@@ -0,0 +1,165 @@
+// NewLanguage returns an instace of the Gazelle plugin for rules_sass.
+func NewLanguage() language.Language {
+	return &sasslang{}
+}
+
+// Kinds returns a map of maps rule names (kinds) and information on how to
+// match and merge attributes that may be found in rules of those kinds. All
+// kinds of rules generated for this language may be found here.
+func (s *sasslang) Kinds() map[string]rule.KindInfo {
+	return map[string]rule.KindInfo{
+		"sass_library": {
+			MatchAny: true,
+			NonEmptyAttrs: map[string]bool{
+				"srcs": true,
+			},
+			MergeableAttrs: map[string]bool{
+				"srcs": true,
+			},
+			ResolveAttrs: map[string]bool{"deps": true},
+		},
+		"sass_binary": {
+			MatchAny: true,
+			NonEmptyAttrs: map[string]bool{
+				"deps": true,
+				"srcs": true,
+			},
+			MergeableAttrs: map[string]bool{
+				"srcs": true,
+				"src":  true,
+			},
+			ResolveAttrs: map[string]bool{"deps": true},
+		},
+	}
+}
+
+// Loads returns .bzl files and symbols they define. Every rule generated by
+// GenerateRules, now or in the past, should be loadable from one of these
+// files.
+func (s *sasslang) Loads() []rule.LoadInfo {
+	return []rule.LoadInfo{
+		{
+			Name:    "@io_bazel_rules_sass//:defs.bzl",
+			Symbols: []string{"sass_binary", "sass_library"},
+		},
+	}
+}
+
+// GenerateRules extracts build metadata from source files in a directory.
+// GenerateRules is called in each directory where an update is requested
+// in depth-first post-order.
+//
+// args contains the arguments for GenerateRules. This is passed as a
+// struct to avoid breaking implementations in the future when new
+// fields are added.
+//
+// empty is a list of empty rules that may be deleted after merge.
+//
+// gen is a list of generated rules that may be updated or added.
+//
+// Any non-fatal errors this function encounters should be logged using
+// log.Print.
+func (s *sasslang) GenerateRules(args language.GenerateArgs) language.GenerateResult {
+	files, err := ioutil.ReadDir(args.Dir)
+	if err != nil {
+		panic(err)
+	}
+
+	// base is the last part of the path for this element. For example:
+	// "hello_world" => "hello_world"
+	// "foo/bar" => "bar"
+	base := filepath.Base(args.Rel)
+
+	rules := []*rule.Rule{}
+	imports := []interface{}{}
+
+	var normalFiles []string
+	for _, file := range files {
+		f := file.Name()
+
+		// Only generate Sass entries for sass files (.scss/.sass)
+		if !strings.HasSuffix(f, ".sass") && !strings.HasSuffix(f, ".scss") {
+			continue
+		}
+
+		fileInfo := sassFileInfo(args.Dir, f)
+		imports = append(imports, fileInfo.Imports)
+
+		// The primary entrypoint on Sass is a main.scss file.
+		if f == "main.scss" {
+			rule := rule.NewRule("sass_binary", base)
+
+			rule.SetAttr("src", "main.scss")
+			rules = append(rules, rule)
+		} else if strings.HasPrefix(filepath.Base(f), "_") {
+			// Libraries in Sass have filenames that start with _.
+			base = filepath.Base(f)
+
+			rule := rule.NewRule("sass_library", base[1:len(base)-5])
```

Comment:
So each `_` file gets its own `sass_library` rules? You'll probably want a way to delete `sass_library` rules after those files have been removed. You can read rules in `args.File` and generate empty rules (returned through `GenerateResult.Empty`) if no rules are generated with those names. The empty rules will be merged, and if there's nothing left, the merged rules will be deleted.

Reply:
Since I only expect there to be a single `sass_library` entry per directory by the name of the directory (if you're in `//foo`, the target would be `//foo:foo`), I've decided to add an empty `rule.NewRule("sass_libary", base)` to the Empty list in all cases. Since Gazelle does a resaonble thing and doesn't delete it if it was generated and deleted this seems like a safe decision. Can you please review this logic.